Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +635,456 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+                return;
+            }
+            m_buffer[m_length + 1 + i] = character;
+        }
+        m_buffer[m_length + 1 + stringLength] = '"';
+        m_length += 1 + stringLength + 1;
+        return;
+    }
+
+    case ObjectType:
+    case FinalObjectType: {
+        auto& object = *asObject(&cell);
+        if (UNLIKELY(object.isCallable())) {
+            recordFailure("callable object"_s);
+            return;
+        }
+        auto& structure = *object.structure();
+        if (UNLIKELY(structure.hasPolyProto())) {
+            recordFailure("hasPolyProto"_s);
+            return;
+        }
+        if (UNLIKELY(structure.storedPrototype() != m_globalObject.objectPrototype())) {
+            recordFailure("non-standard object prototype"_s);
+            return;
+        }
+        if (!m_checkedObjectPrototype) {
+            if (UNLIKELY(mayHaveToJSON(*m_globalObject.objectPrototype()))) {
+                recordFailure("object prototype may have toJSON"_s);
+                return;
+            }
+            m_checkedObjectPrototype = true;
+        }
+        if (UNLIKELY(!remainingCapacity())) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length++] = '{';
+        if (UNLIKELY(!canPerformFastPropertyEnumerationForJSONStringify(&structure))) {
+            recordFastPropertyEnumerationFailure(object);
+            return;
+        }
+        structure.forEachProperty(m_vm, [&](const PropertyTableEntry& entry) -> bool {
+            if (entry.attributes() & PropertyAttribute::DontEnum)
+                return true;
+            auto& name = *entry.key();
+            if (UNLIKELY(name.isSymbol())) {
+                recordFailure("symbol"_s);
+                return false;
+            }
+            if (UNLIKELY(!name.is8Bit())) {
+                recordFailure("16-bit property name"_s);
+                return false;
+            }
+            bool needComma = m_buffer[m_length - 1] != '{';
+            unsigned nameLength = name.length();
+            if (UNLIKELY(remainingCapacity() < needComma + 1 + nameLength + 2)) {
+                recordBufferFull();
+                return false;
+            }
+            if (needComma)
+                m_buffer[m_length++] = ',';
+            m_buffer[m_length] = '"';
+            auto* characters = name.characters8();
+            for (unsigned i = 0; i < nameLength; ++i) {
+                auto character = characters[i];
+                if (UNLIKELY(WTF::escapedFormsForJSON[character])) {
+                    recordFailure("property name character needs escaping"_s);
+                    return false;
+                }
+                m_buffer[m_length + 1 + i] = character;
+            }
+            m_buffer[m_length + 1 + nameLength] = '"';
+            m_buffer[m_length + 1 + nameLength + 1] = ':';
+            m_length += 1 + nameLength + 2;
+            append(object.getDirect(entry.offset()));
```

Comment:
I think still it is possible that non-user-observable behavior can change the structure. Like, reify of properties, custom values etc.
So, let's add a check,
```
if (UNLIKELY(&structure != object.structure()))
    // FAIL
```
Even though any kind of side-effect happens, if structure is not changed, then it is OK.

Reply:
I am **absolutely** willing to add this check if it’s needed; I will need to re-run performance testing, but I bet it won’t hurt too much. But I still don’t understand. Is it `getDirect` that can reify properties or `forEachProperty`? Can I avoid this by using `propertyTableOrNull` instead?

Reply:
No, what I'm worrying is some operations in `append(...)`. Do we have any guarantee that some code in `append` will not invoke the structure transition?

Reply:
This function itself is `append`; it’s a recursive call to this function. We can look over the whole function to see what is risky and could cause a structure transition. To me it seems obvious that nothing `append` does can cause a transition, but maybe that’s because I don’t know what to look for.

Reply:
By the way, it would be OK if it causes a transition and a failure, because if we have a failure it immediately exits. If it can cause a transition and succeed then, yes, that would it be a problem.

As I said, I have the humility to admit there could be a transition because my understanding isn’t deep enough to appreciate something we are doing here in `append` that could have that subtle effect, but I **don’t** think there’s anything here that can do that.

Reply:
If we missed it, it becomes serious security issue since it allows attackers to access to non-used place with newly transitioned (sometimes, it gets flattened) structure. Historically, there are many missed cases (e.g. Object.assign's reify case). And since structure transition is not user-observable behavior, there is no guarantee that future code will perform structure transition, which is used in somewhere in `append` function, and we cannot notice that this place becomes a security issue after that.
So, so long as we have 100% confident & we have a mechanism to assert this guarantee, I think we should check it.
Compared to user-observable-behaviors, JSC developers do not see structure transitions as a side-effect in the code, so I cannot guarantee that `append` can / will not include that code.

Reply:
OK, I can put in an assertion and a check from a general sense of security vulnerability paranoia/hardening

Reply:
Is the pointer check sufficient? Can the old structure’s memory be deallocated as part of a structure transition and possible be reused for a new structure?

Reply:
New patch uploaded has this check added.